keps/sig-cluster-lifecycle/kubeadm/20191229-kubeadm-move-out-of-kk.md

```diff
@@ -0,0 +1,635 @@
+the current version of Go that the k/k repository uses. Syncing Go versions
+should be the responsibility of the k/kubeadm maintainers.
+
+The new PR blocking jobs must support building kubeadm from source with
+the latest commits that a PR for k/kubeadm introduces. The rest of the
+k8s artifacts can be pulled from the CI GCS bucket. This can also allow
+catching problems introduced in k/k when using kubeadm as the deployer.
+
+The /hack/verify-* and update-* scripts that k/kubeadm/kubeadm uses
+will not be a direct copy of k/k but will provide similar results:
+Go versions are in sync Go-fmt, linter and import tools are in sync
+as much as possible
+
+The following lists of tests will be executed on PRs as blocking:
+- Unit tests
+- Integration tests (currently reside in k/k/cmd/kubeadm/test)
+- E2e tests from the e2e-kubeadm suite (currently resides
+in k/k/test/e2e-kubeadm)
+- Full cluster deployment e2e, possibly upgrade and skew tests too
+
+#### Periodics jobs
+
+No changes to the current kubeadm periodic jobs will be required
+as long as the CI build process continues to push the kubeadm
+and the rest of the k8s binaries to the GCS buckets.
+
+### Graduation Criteria
+
+Not applicable as the work in this KEP should be ideally executed within
+one k8s release cycle.
+
+### Upgrade / Downgrade Strategy
+
+No changes to the upgrade strategy will be required.
+Downgrades are not supported by kubeadm.
+
+### Version Skew Strategy
+
+No changes will be applied to the current version skew support in kubeadm
+as long as kubeadm preserves its exiting versioning, which matches
+the versioning of k8s.
+
+## Implementation History
+
+- 2019-12-29: This proposal document was created
+
+## Open Questions
+
+#### Can we stop building DEBs and RPMs from k/k?
+
+Currently the Kubernetes CI build process uses Bazel specs from the
+k/k `/build` directory to build DEBs and RPMs. Packages are generated
+for kubelet, kubeadm as the main targets. The Bazel specs assume kubeadm
+exists under `/cmd/kubeadm` so this process has to be either changed
+or completely removed.
+
+Arguments for the removal of this process from k/k:
+
+- The existence of the CI artifacts is not documented in kubeadm
+and kubelet documentation.
+- Observations by triaging issues in k/k and k/kubeadm in the past
+couple of years have revealed no users of the CI DEBs and RPMs.
+- DEBs and RPMs are also managed in k/release thus the k/k package
+specs can end up out of date.
+
+This KEP proposes the that k/k DEB and RPM build process is completely
+removed from k/k and that it is consolidated over at the k/release
+repository, where there is ongoing work to refactor the package building.
+
+## Drawbacks
+
+NONE
+
+## Alternatives
+
+A number of proposals for automating parts of the kubeadm maintenance
+process once it moves out have been made. Such improvements are considered
+nice-to-have and are not blocking for the initial work of moving kubeadm.
```

Comment:
I would like to see listed as alternative: "Running k/kubeadm independent of k/k, with its own branching and versioning." and the pros/cons of that given some discussion.  In reading the KEP the only con I sense is that kubeadm's version string would not be the same as k/k's version string and users needing to understand the intended support matrix between kubeadm versions and k/k versions.

Reply:
this con itself is enough of a problem. some users are confused nowadays about the current skew. if introduce the custom versioning this will create quite the confusion.

having the versioning to match k/k is a preference of the maintainers.
this is a core aspect of the proposal and i don't think we should outline the alternative.

it is not clear to me how kubectl will handle their decoupling from the k8s versioning. from a brief discussion with them i don't think it's clear to them either.

Reply:
ok, i'm going to add some comments here related to alternatives related to the post-submit job (prow plugin) and alternatives related to the process of auto-branching and tagging (e.g. manual).